frontend/src/components/teacher/student-management/classroom-summary/AddOrEditClassroomModal.tsx

```diff
@@ -83,168 +67,111 @@ const AddOrEditClassroomModal = ({
-  const handleDateChange = (date: Date) => {
-    setValue("startDate", date, { shouldDirty: true });
-  };
-
-  const validateFields = (): boolean => {
-    if (!watch("className") || !!errors.className) {
-      return false;
-    }
-
-    if (!watch("startDate") || !!errors.startDate) {
-      return false;
-    }
-
-    if (!watch("gradeLevel") || !!errors.gradeLevel) {
-      return false;
-    }
-    return true;
-  };
+  const upsertClass = classroomId ? updateClass : createClass;
 
   const onModalClose = () => {
     resetForm();
-    setShowRequestError(false);
-    setRequestErrorMessage("");
     onClose();
   };
 
-  const onSave: SubmitHandler<ClassroomForm> = async (data) => {
-    const startDate = watch("startDate");
-    const now = new Date();
-
-    if (!validateFields()) {
-      setShowRequestError(true);
-      setRequestErrorMessage(
-        "Please ensure all required components are filled out before saving changes",
-      );
-    } else if (
-      dirtyFields.startDate &&
-      startDate &&
-      startDate < now &&
-      !isSameDay(startDate, now)
-    ) {
-      setShowRequestError(true);
-      setRequestErrorMessage("Please set a present or future date");
-    } else {
-      if (showRequestError) setShowRequestError(false);
-      const classObj = {
-        ...data,
-        startDate: data.startDate,
-        teacher: authenticatedUser?.id,
-      };
+  const onSave = handleSubmit((data) =>
+    upsertClass({
+      variables: {
+        classroomId,
+        classObj: {
+          ...data,
+          teacher: authenticatedUser?.id,
+        },
+      },
+    }),
+  );
 
-      if (classroomId) {
-        try {
-          await updateClass({
-            variables: {
-              classroomId,
-              classObj,
-            },
-          });
-          history.replace(history.location.pathname, undefined);
-          showToast({
-            message: "Classroom information updated.",
-            status: "success",
-          });
-        } catch (e) {
-          showToast({
-            message:
-              "Failed to update the classroom's information. Please try again.",
-            status: "error",
-          });
-        }
-      } else {
-        try {
-          await createClass({
-            variables: {
-              classObj: {
-                ...data,
-                startDate: data.startDate,
-                teacher: authenticatedUser?.id,
-              },
-            },
-          });
-          showToast({
-            message: "New classroom created.",
-            status: "success",
-          });
-        } catch (e) {
-          showToast({
-            message: "Failed to create a new classroom. Please try again.",
-            status: "error",
-          });
-        }
-      }
-      onModalClose();
+  const handleSave = async () => {
+    await onSave();
+    if (!formState.isValid) {
+      throw new FormValidationError("Please ensure all fields are valid");
     }
```

Comment:
```suggestion
    if (!formState.isValid) {
      throw new FormValidationError("Please ensure all fields are valid");
    }
    await onSave();
```
?

Reply:
The order actually matters here because validation occurs in what is called `onSave` here. I updated the naming to be more clear and extracted a custom hook for reuse elsewhere to hide the weird-ish logic.